mat64/io.go

```diff
@@ -5,14 +5,180 @@
 package mat64
 
 import (
+	"bytes"
 	"encoding/binary"
+	"errors"
+
+	"github.com/gonum/matrix"
 )
 
-var (
-	littleEndian  = binary.LittleEndian
-	bigEndian     = binary.BigEndian
-	defaultEndian = littleEndian
+// maxLen is the biggest slice/array len one can create on a 32/64b platform.
+const maxLen = int64(int(^uint(0) >> 1))
 
+var (
 	sizeInt64   = binary.Size(int64(0))
 	sizeFloat64 = binary.Size(float64(0))
+
+	errTooBig    = errors.New("mat64: resulting data slice too big")
+	errBadBuffer = errors.New("mat64: data buffer size mismatch")
 )
+
+// MarshalBinary encodes the receiver into a binary form and returns the result.
+//
+// Dense is little-endian encoded as follows:
+//   0 -  7  number of rows    (int64)
+//   8 - 15  number of columns (int64)
+//  16 - ..  matrix data elements (float64)
+//           [0,0] [0,1] ... [0,ncols-1]
+//           [1,0] [1,1] ... [1,ncols-1]
+//           ...
+//           [nrows-1,0] ... [nrows-1,ncols-1]
+func (m Dense) MarshalBinary() ([]byte, error) {
+	bufLen := int64(m.mat.Rows*m.mat.Cols*sizeFloat64 + 2*sizeInt64)
+	if bufLen > maxLen {
```

Comment:
> It is a failure for when a contiguous allocation cannot be made, while this failure here would be for spuriously large size descriptors.

Isn't it the same? We care about large size descriptors only because we need to allocate the destination buffer. As long as we can get the buffer and write to it, Marshal does what it is supposed to do. In other words, what is the purpose of this check? To avoid panic in `make` below? Does this check guarantee that it won't panic?